### PR TITLE
Various matrix fixes

### PR DESCRIPTION
**Pull request by jdevnull**
(snipped)
Start here: https://github.com/luciensadi/AwakeMUD/pull/735#issuecomment-2126371998

### CODE REVIEW

**Comment by jdevnull** [comment]
Colus' statement at [https://discord.com/channels/564618629467996170/564621949842751508/1217174615449927842](url) that trap doors should drop you directly into the target node makes sense to me, so here's the code for it.

After [https://discord.com/channels/564618629467996170/1216675006164242432](url), what I'm thinking is to give each host a linked list of entrances (similar to how they each have a linked list of exits), and to populate the lists at the time of load. But I can't seem to find where in the code the hosts are loaded from the database....

**Comment by luciensadi** [comment]
Game world loading is in db.cpp, with various loaders called from `boot_world()`. If you follow `DB_BOOT_MTX` through, you'll eventually hit `discrete_load()`, which sends you to `parse_host()`, which is the actual host loading function for individual hosts. In this case though, you'd probably just want to stick your list-building code _after_ the world loads, probably right around the commented-out "Creating Help Indexes" line or even after `boot_world()` itself is called.

While I won't stop you from adding a linked list of entrances, remember that if that list is full of pointers, you're going to have to update every list every time a host is moved around in memory or an exit is created/deleted/etc. I recommend testing under Valgrind if you're adding something like that since it will have a higher chance of catching mistakes (it regularly catches mine when I do things along those lines since the code is full of spaghetti).

**Comment by jdevnull** [comment]
In newmatrix.cpp, `matrix_update()` starts from `matrix[1]`, but db.cpp functions start from `matrix[0]`. Is `matrix[0]` an actual host?

**Comment by luciensadi** [comment]
> In newmatrix.cpp, `matrix_update()` starts from `matrix[1]`, but db.cpp functions start from `matrix[0]`. Is `matrix[0]` an actual host?

No. It exists, but it's host vnum 0 which is not used for anything in the game. 1-indexing is *teeeeechnically* a bug, but it has no game impact and might be a workaround for a crash related to updating a host with a ton of nulled fields, so I'd leave it be.

**Comment by jdevnull** [comment]
Notes on current PR:
1. When attempting to logon to a local subsystem that has a trapdoor, the original code immediately sets the host for the system test to the destination host. The end result was that the _local_ subsystem's encryption actually _doesn't matter_ (in addition to the problem that there's no way to decrypt the destination host's access through a backdoor). Current PR will check the local subsystem first for encryption, forwards to the destination host if it's not encrypted, and ignores the destination host's access encryption because these are supposed to be back doors.
2. Current PR also provides each host with a linked list of entrances, which is reset and re-populated from every hosts' linked list of exits (via `collate_host_entrances()`) after `boot_world()` (db.cpp) loads all the matrix hosts, and again whenever host changes are saved to disk (hedit.cpp). These entrance lists are then used to check for decker presence when we want to re-encrypt hosts in `matrix_update()` (newmatrix.cpp).

Related to that, every time `matrix_update()` is called (which is every 2 IRL seconds), if time_info.hours == 0, then we check to re-encrypt each and every host that doesn't have a decker in the area. So this means that the re-encryption code is executed 60 times during the hour of midnight. I'm thinking of changing this to something like checking only once (using some kind of flag similar to payreset) but with an interval of every 6 hours, but maybe what's already here should be reviewed and tested first.

**Comment by luciensadi** [comment]
Thanks for submitting! I'm behind on reviewing PRs, but will try to take a look this week. Is this ready for review or still in progress?

**Comment by jdevnull** [comment]
Ready!

**Comment by luciensadi** [comment]
The change here is pretty complex, and I _think_ I have my head around it but am not positive. Did this all work as expected when you tested it?

**Comment by jdevnull** [comment]
Well, the code chunks seem to do what I want them to do when I tested them separately, but I haven't tested it integrated into the mud. Difficulty is that my only real linux machine at the moment is my work computer, so I'm a bit limited on what I can install on there.

An option might be installing it on the WSL side of my win11 laptop... hmm.

**Comment by jdevnull** [comment]
Missed some interactions. When attempting to logon to a subsystem, the code only checks the ACCESS subsystem for encryption instead of where the trap door actually is.

**Comment by jdevnull** [comment]
Alright, revised how I approached handling trap doors, so here's the summary of what this PR does:
1. Each host now has a linked list of entrances. If host A has an exit to host B, if host B is A's parent (so that logon LTG/RTG in A sends you to B), or if A has a subsystem trap door that leads to B, then A is an entrance to B. These lists are re-populated on boot and when saving builder matrix changes.
2. So long as there is no decker in host B or at any of the entrances to B, then B's subsystems are permitted to re-encrypt.
3. Separated out the matrix updates that should be checked no more than once an hour into `matrix_hour_update()`. This way we're not checking all the subsystems for re-encryption every 2 seconds during whichever mud hour host sysops are reviewing security. We also no longer need the payreset lock.
4. System tests against trap doors target the _local subsystem_ instead of the destination host. This allows the scope of the `system_test` function to remain as it was while making the following fixes: (1) encryption of the local subsystem now needs to be decrypted before using the trap door, and (2) encryption of the destination host's access no longer makes the trap door impenetrable.
5. Added host alert state decay (active to passive, passive to no alert), sharing the same requirements as for subsystem re-encryption. This is checked every 2 mud hours so that alert decay has a similar timescale as physical mob alert/alarm states.

Tested using The Cave / Waterfall hosts in Portland. Did NOT test builder actions (since I'm entirely unfamiliar with being a builder).

This should address the following bug reports:
https://discord.com/channels/564618629467996170/1208653573676597248
https://discord.com/channels/564618629467996170/1209262641349005363
https://discord.com/channels/564618629467996170/1219900280926240821
https://discord.com/channels/564618629467996170/1218574342565138484
https://discord.com/channels/564618629467996170/1242383821022826498
(and others)

File: src/newmatrix.cpp [hunk 1]
```diff
@@ -1738,34 +1735,40 @@ ACMD(do_logon)
...
-    target_host = real_host(matrix[PERSONA->in_host].stats[INDEX][MTX_STAT_TRAPDOOR]);
-  else if (!str_cmp(argument, "files") && matrix[PERSONA->in_host].stats[FILES][MTX_STAT_TRAPDOOR])
-    target_host = real_host(matrix[PERSONA->in_host].stats[FILES][MTX_STAT_TRAPDOOR]);
-  else if (!str_cmp(argument, "slave") && matrix[PERSONA->in_host].stats[SLAVE][MTX_STAT_TRAPDOOR])
-    target_host = real_host(matrix[PERSONA->in_host].stats[SLAVE][MTX_STAT_TRAPDOOR]);
-  else {
-    for (struct exit_data *exit = matrix[PERSONA->in_host].exit; exit; exit = exit->next)
+  } else {
+    // Subsystem trap door? Test against the local subsystem
+    for (int sub = ACCESS; (target_host < 0) && (sub < NUM_OF_SUBSYSTEMS); sub++) {
```
Comment (luciensadi) [inline src/newmatrix.cpp:1740]: What's the purpose of the `target_host < 0` block? If it's to terminate iteration on a successful if-check, this is better done with an explicit `break` statement.

File: src/newmatrix.cpp [hunk 2]
```diff
@@ -3110,6 +3065,75 @@ void matrix_update()
...
+
+        // Re-encrypt subsystems.
+        for (int x = ACCESS; x < NUM_OF_SUBSYSTEMS; x++) {
+          if (HOST.stats[x][MTX_STAT_ENCRYPTED] != 0 && HOST.stats[x][MTX_STAT_ENCRYPTED] != 1) {
+            char warnbuf[1000];
+            snprintf(warnbuf, sizeof(warnbuf), "WARNING: %s mtx_stat_encrypted on %ld is %ld (must be 1 or 0)!",
+                      acifs_strings[x],
+                      HOST.vnum,
+                      HOST.stats[x][MTX_STAT_ENCRYPTED]);
+            mudlog(warnbuf, NULL, LOG_SYSLOG, TRUE);
```
Comment (luciensadi) [inline src/newmatrix.cpp:3123]: No need to change, but FYI this is old-style logging. You can now use `mudlog_vfprintf(NULL, LOG_SYSLOG, "message", var1, var2 ...);` and skip the declaration of a buffer and sprintf() call.

File: src/newmatrix.cpp [hunk 3]
```diff
@@ -3110,6 +3065,75 @@ void matrix_update()
   }
 }
 
+// For things that should happen no more than once an hour.
+void matrix_hour_update()
+{
+  PERF_PROF_SCOPE(pr_, __func__);
+  rnum_t rnum = 1;
```
Comment (luciensadi) [inline src/newmatrix.cpp:3072]: Since this isn't used outside of the for-loop, please define it in the loop initializer (`for (rnum_t rnum = 1; ...)`)

**Review by luciensadi — COMMENTED** [review]
Thanks for submitting! Added a few thoughts.

File: src/newmatrix.cpp [hunk 4]
```diff
@@ -1724,7 +1719,9 @@ ACMD(do_logon)
     return;
   }
   skip_spaces(&argument);
-  rnum_t target_host = 0;
+  rnum_t target_host = -1, trapdoor_host = -1;
+  int subsystem = ACCESS;  // Logon usually targets ACCESS 
+
   WAIT_STATE(ch, (int) (DECKING_WAIT_STATE_TIME));
   if (!str_cmp(argument, "LTG")) {
     if (!(target_host = real_host(matrix[PERSONA->in_host].parent))
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp:1727]: Huh. That's been the code for a while, I'm surprised this hasn't been crashing the game.

File: src/newmatrix.cpp (hunk 3 above)
Comment (jdevnull) [reply src/newmatrix.cpp:3072]: Yeah, this was a copy/paste from the existing `matrix_update()`, didn't really think about it.

File: src/newmatrix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp:1727]: Reading into the next line, I think there's actually two bugs. I believe the logic is supposed to return the message if parent is not real, or parent is not an LTG/PLTG.

File: src/db.cpp [hunk 5]
```diff
@@ -7579,3 +7600,54 @@ void set_elemental_races() {
...
+      entrance = new entrance_data;
+      entrance->host = &matrix[source_rnum];
+      entrance->next = matrix[dest_rnum].entrance;
+      matrix[dest_rnum].entrance = entrance;
+    }
+
+    // Subsystem trapdoors are also entrances
+    for (int sub = ACCESS; sub < NUM_OF_SUBSYSTEMS; sub++) {
+      dest_rnum = real_host(matrix[source_rnum].stats[sub][MTX_STAT_TRAPDOOR]);
+      if (dest_rnum >= 0) {
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/db.cpp]: Done

File: src/db.cpp [hunk 6]
```diff
@@ -7579,3 +7600,54 @@ void set_elemental_races() {
...
+    for (entrance = matrix[source_rnum].entrance; entrance; entrance = next) {
+      next = entrance->next;
+      delete entrance;
+    }
+    matrix[source_rnum].entrance = NULL;
+  }
+
+  // Find exits and add to respective entrance lists
+  rnum_t dest_rnum = -1;
+  struct entrance_data *entrance = NULL;
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/db.cpp]: Done

File: src/newmatrix.cpp [hunk 7]
```diff
@@ -1237,7 +1232,7 @@ ACMD(do_locate)
   two_arguments(argument, buf, arg);
   int success, i = 0;
   if (is_abbrev(buf, "hosts")) {
-    success = system_test(PERSONA->in_host, ch, TEST_INDEX, SOFT_BROWSE, 0);
+    success = system_test(PERSONA->in_host, ch, INDEX, SOFT_BROWSE, 0);
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp]: Done

File: src/newmatrix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp]: Partially fixed (comments added to code).

File: src/newmatrix.cpp (hunk 1 above)
Comment (jdevnull) [reply src/newmatrix.cpp]: Done

File: src/newmatrix.cpp (hunk 3 above)
Comment (jdevnull) [reply src/newmatrix.cpp]: Done

**Comment by jdevnull** [comment]
And a few more changes:
6. Corrected calculation of remaining hacking pool when resisting an attack (esp. so that it doesn't look like a cold ASIST is using pool dice).
7. Running the medic utility can use hacking pool dice (SR3 pg 207).
8. Only one armor program (the first found) applies to damage resistance.

**Comment by luciensadi** [comment]
LGTM, thanks for all your hard work here!

**Comment by luciensadi** [comment]
FYI the linked list of entrances causes invalid memory access when editing an existing host and 'X'ing out. I think it's trying to clear entrances related to the editing struct and failing when it finds none?

**Comment by jdevnull** [comment]
Hmm. I bet this was from what I did to `free_host`. I'm thinking the search/removal of entrances in `free_host` isn't actually necessary since `collate_host_entrances` completely clears and rebuilds the entrance lists anyway.

**Pull request merged**